clang/runtime/dpct-rt/include/sparse_utils.hpp

```diff
@@ -743,6 +767,316 @@ inline void spmm(sycl::queue queue, oneapi::mkl::transpose trans_a,
+/// \param [in] a Specifies the sparse matrix a.
+/// \param [in] b Specifies the sparse matrix b.
+/// \param [in] beta Specifies the scalar beta.
+/// \param [in, out] c Specifies the sparse matrix c.
+/// \param [in] matmat_descr Describes the sparse matrix-sparse matrix operation
+/// to be executed.
+/// \param [in, out] size_temp_buffer Specifies the size of workspace.
+/// \param [in] temp_buffer Specifies the memory of the workspace.
+inline void spgemm_compute(sycl::queue queue, oneapi::mkl::transpose trans_a,
+                           oneapi::mkl::transpose trans_b, const void *alpha,
+                           sparse_matrix_desc_t a, sparse_matrix_desc_t b,
+                           const void *beta, sparse_matrix_desc_t c,
+                           oneapi::mkl::sparse::matmat_descr_t matmat_descr,
+                           size_t *size_temp_buffer, void *temp_buffer) {
+  set_matmat_data(matmat_descr, oneapi::mkl::sparse::matrix_view_descr::general,
+                  trans_a, oneapi::mkl::sparse::matrix_view_descr::general,
+                  trans_b, oneapi::mkl::sparse::matrix_view_descr::general);
+  if (temp_buffer) {
+#ifdef DPCT_USM_LEVEL_NONE
+    sycl::buffer<std::uint8_t> extBuf =
+        dpct::get_buffer<std::uint8_t>(temp_buffer);
+    sycl::buffer<std::int64_t, 1> bufSizeBuf(sycl::range<1>(1));
+    auto bufSizeBuf_acc = bufSizeBuf.get_host_access(sycl::write_only);
+    bufSizeBuf_acc[0] = static_cast<std::int64_t>(*size_temp_buffer);
+    oneapi::mkl::sparse::matmat(queue, a->get_matrix_handle(),
+                                b->get_matrix_handle(), c->get_matrix_handle(),
+                                oneapi::mkl::sparse::matmat_request::compute,
+                                matmat_descr, &bufSizeBuf, &extBuf);
+    std::int64_t nnzValue = 0;
+    {
+      sycl::buffer<std::int64_t, 1> nnzBuf(&nnzValue, sycl::range<1>(1));
+      oneapi::mkl::sparse::matmat(
+          queue, a->get_matrix_handle(), b->get_matrix_handle(),
+          c->get_matrix_handle(), oneapi::mkl::sparse::matmat_request::get_nnz,
+          matmat_descr, &nnzBuf, nullptr);
+    }
+    c->set_nnz(nnzValue);
+#else
+    std::int64_t *bufSize = sycl::malloc_host<std::int64_t>(1, queue);
+    *bufSize = static_cast<std::int64_t>(*size_temp_buffer);
+    oneapi::mkl::sparse::matmat(queue, a->get_matrix_handle(),
+                                b->get_matrix_handle(), c->get_matrix_handle(),
+                                oneapi::mkl::sparse::matmat_request::compute,
+                                matmat_descr, bufSize, temp_buffer, {});
+    std::int64_t *c_nnz = sycl::malloc_host<std::int64_t>(1, queue);
+    oneapi::mkl::sparse::matmat(queue, a->get_matrix_handle(),
+                                b->get_matrix_handle(), c->get_matrix_handle(),
+                                oneapi::mkl::sparse::matmat_request::get_nnz,
+                                matmat_descr, c_nnz, nullptr, {});
```

Comment:
Your USM APIs are problematic. You are not keeping track of and carrying over dependencies. This won't work. Please fix this (extract the event, `ev_compute = matmat(...)`, then `ev_nnz = matmat(..., {ev_compute})`, etc.)

Reply:
The queue used here should be an in-ordered queue. So I think the explicit dependency setting is not necessary.

Reply:
Isn't the queue provided to you by the user? Is the queue under the complete control of DPCT? Doesn't the user have the freedom to change the queue to be out-of-order on their end once DPCT tool ports over CUDA code?

Reply:
If the code is totally migrated by DPCT, the tool will maintain all queues and make sure queues are all in-ordered (in USM case).
If user manually changes the migrated code, user needs to make sure the input queue of this API is in-ordered. We have added document for this parameter above this API.